compiler/rustc_parse/src/parser/path.rs

```diff
@@ -675,14 +675,39 @@ impl<'a> Parser<'a> {
             GenericArg::Const(self.parse_const_arg()?)
         } else if self.check_type() {
             // Parse type argument.
-            let is_const_fn =
-                self.look_ahead(1, |t| t.kind == token::OpenDelim(Delimiter::Parenthesis));
-            let mut snapshot = self.create_snapshot_for_diagnostic();
-            match self.parse_ty() {
-                Ok(ty) => GenericArg::Type(ty),
+
+            let mut snapshot = None;
```

Comment:
This parsing logic has gotten complex enough that it probably deserves a comment block explaining what it's trying to do 🤔

Reply:
I can also try to factor out the recovery logic into a separate function, come to think of it.

Reply:
I've added some comments. Let me know if they are helpful or confusing.